### PR TITLE
Ftr 2198 timed notifications

### LINKED ISSUE #2195: Notification: no course assessment request
As a student of Nettilukio i would like to receive a communicator message and email when i have not sent any course assessment requests in 2 months and 2,5 months.

### LINKED ISSUE #2196: Notification: no passed courses
As a student of Nettilukio i would like to receive a communicator message and an email when i have not passed at least 5 courses in last 10 months.

### LINKED ISSUE #2197: Notification: study time left
As a student of Nettilukio or Internetix i would like to receive a communicator message and an email when I have only a month of study time left.

### LINKED ISSUE #2198: Notification: not added requested assessment supplementations
As a student of Nettilukio i would like to receive a communicator message and an email when i have not added requested assessment supplementations within one week.

### PR DESCRIPTION
**Pull request by hkurhinen**
fixes #2195 
fixes #2196 
fixes #2197 
fixes #2198

### CODE REVIEW

File: muikku-core/src/main/java/fi/otavanopisto/muikku/schooldata/GradingController.java [hunk 1]
```diff
@@ -173,6 +174,25 @@ public WorkspaceAssessmentRequest findWorkspaceAssessmentRequest(String schoolDa
...
+    } catch (SchoolDataBridgeRequestException | UnexpectedSchoolDataBridgeException e) {
+      logger.log(Level.SEVERE, String.format("Failed to list student assessment requests for student %s", studentIdentifier.toId()), e);
+    }
+    return Collections.emptyList();
+  }
+  
+  public List<WorkspaceAssessmentRequest> listStudentAssessmentRequestsSince(SchoolDataIdentifier studentIdentifier, Date date) {
+    List<WorkspaceAssessmentRequest> result = new ArrayList<>();
+    for (WorkspaceAssessmentRequest workspaceAssessmentRequest : listStudentAssessmentRequests(studentIdentifier)) {
+      if (workspaceAssessmentRequest.getDate().after(date)) {
```
Comment (anttiviljakainen) [inline muikku-core/src/main/java/fi/otavanopisto/muikku/schooldata/GradingController.java]: getDate() nullcheck?

File: muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java [hunk 2]
```diff
@@ -130,12 +132,28 @@ public SearchResult searchUsers(String text, String[] textFields, Collection<Env
         }
       }
       
+      if (excludeSchoolDataIdentifiers != null) {
+        IdsFilterBuilder excludeIdsFilterBuilder = FilterBuilders.idsFilter("User");
+        for (SchoolDataIdentifier excludeSchoolDataIdentifier : excludeSchoolDataIdentifiers) {
+          excludeIdsFilterBuilder.addIds(String.format("%s/%s", excludeSchoolDataIdentifier.getIdentifier(), excludeSchoolDataIdentifier.getDataSource()));
```
Comment (anttiviljakainen) [inline muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java]: is there another (more generic) way to do this without string.format or is it handled like this in the source?

Comment (hkurhinen) [inline muikku-elastic-search/src/main/java/fi/otavanopisto/muikku/plugins/search/ElasticSearchProvider.java]: Ids are (for some reason?) indexed like this, there is method toString in SchoolDataIdentifier but it returns a string where identifier and datasource are in wrong order.

Similar method was used in already existing "includeIds" filter:
```java
if (userIdentifiers != null) {
  IdsFilterBuilder idsFilter = FilterBuilders.idsFilter("User");
   for (SchoolDataIdentifier userIdentifier : userIdentifiers) {
     idsFilter.addIds(String.format("%s/%s", userIdentifier.getIdentifier(), userIdentifier.getDataSource()));
   }
   filters.add(idsFilter);
}

```

File: muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/RequestedAssessmentSupplementationsNotificationStrategy.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            if (assessment.getDate().before(new DateTime().minusDays(NOTIFICATION_THRESHOLD_DAYS).toDate())) {
+              GradingScale gradingScale = gradingController.findGradingScale(assessment.getGradingScaleSchoolDataSource(), assessment.getGradingScaleIdentifier());
+              GradingScaleItem grade = gradingController.findGradingScaleItem(gradingScale, assessment.getGradeSchoolDataSource(), assessment.getGradeIdentifier());
+              
+              if (!grade.isPassingGrade()) {  
+                try {
+                  WorkspaceAssessmentRequest latestAssesmentRequest = null;
+                  List<WorkspaceAssessmentRequest> studentAssesmentRequests = gradingController.listWorkspaceAssessmentRequests(workspaceIdentifier.getDataSource(), workspaceIdentifier.getIdentifier(), studentIdentifier.getIdentifier());
+                  for (WorkspaceAssessmentRequest assessmentRequest : studentAssesmentRequests) {
+                    if (latestAssesmentRequest == null || latestAssesmentRequest.getDate().before(assessmentRequest.getDate())) {
```
Comment (anttiviljakainen) [inline muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/RequestedAssessmentSupplementationsNotificationStrategy.java]: getDate null?

File: muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/RequestedAssessmentSupplementationsNotificationStrategy.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                      latestAssesmentRequest = assessmentRequest;
+                    }
+                  }
+                  if (latestAssesmentRequest == null || latestAssesmentRequest.getDate().before(assessment.getDate())) {
```
Comment (anttiviljakainen) [inline muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/RequestedAssessmentSupplementationsNotificationStrategy.java]: getDate null?

File: muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/StudyTimeNotificationStrategy.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    List<SchoolDataIdentifier> studentIdentifiers = new ArrayList<>();
+    for (Map<String, Object> result : searchResult.getResults()) {
+      String studentId = (String) result.get("id");
+      
+      if (StringUtils.isBlank(studentId)) {
+        logger.severe("Could not process user found from search index because it had a null id");
+        continue;
+      }
+      
+      String[] studentIdParts = studentId.split("/", 2);
```
Comment (anttiviljakainen) [inline muikku-timed-notifications-plugin/src/main/java/fi/otavanopisto/muikku/plugins/timed/notifications/strategies/StudyTimeNotificationStrategy.java]: Generic function (SchoolDataIdentifier..)?

**Pull request closed without merging**